### PR TITLE
#1037 Delete outdated docker images

### LINKED ISSUE #1037: After a feature/bug/patch/hotfix has been merged, the respective (temporary) images should be deleted
**Problem:** At the moment, all image builds are stored on DockerHub, e.g., the installer image ( https://hub.docker.com/r/keptn/installer/tags ) has 39 pages of tags (roughly 963 tags).

**To find old images:** Old images for feature/bug/patch/hotfix branches can be found using the following command:
```console
wget -q https://registry.hub.docker.com/v1/repositories/keptn/installer/tags -O -  | sed -e 's/[][]//g' -e 's/"//g' -e 's/ //g' | tr '}' '\n'  | awk -F: '{print $3}' | grep "patch\|feature\|hotfix\|bug"
```

To delete tags of an image, use the example provided in https://github.com/docker/hub-feedback/issues/496#issuecomment-277562292 

**Tasks:**
* Implementation of Travis-CI script / GH Action that : 
  * Runs once per week and only delete tags that are older than 1 month.
  * Deletes "dirty" tags, which will be produced with in-cluster debugging.

**Additional Info: Dirty images**
Get all dirty tags of `keptn/helm-service`: https://hub.docker.com/repository/registry-1.docker.io/keptn/helm-service/tags?page=1&name=dirty

Programmatically
```console
TOKEN=$(curl -s -H "Content-Type: application/json" -X POST -d '{"username": "'${DOCKER_HUB_USERNAME}'", "password": "'${DOCKER_HUB_PASSWORD}'"}' https://hub.docker.com/v2/users/login/ | jq -r .token)

curl -s -H "Authorization: JWT ${TOKEN}" "https://hub.docker.com/v2/repositories/keptn/helm-service/tags/?name=dirty&page_size=100"  | jq -r '.results|.[]|.name'
```
Output:
```
0.6.1-774-g2ad90513-dirty
0.6.1-637-g31684110-dirty
0.6.1-635-ga898007a-dirty
0.6.1-602-g4ca43d85-dirty
0.6.1-588-gf2ff0dbb-dirty
0.6.1-586-g913742f6-dirty
0.6.1-534-gf9eb7317-dirty
0.6.1-499-g0ec81d51-dirty
0.6.1-500-gf720f2c6-dirty
0.6.1-411-gca02f275-dirty
0.6.1-404-g46b74cf8-dirty
0.6.1-271-ga1cd1630-dirty
0.6.1-256-g9ab928f3-dirty
0.6.1-1-g53c9b892-dirty
0.6.1-152-g12f0637f-dirty
0.6.1-147-g9de1451d-dirty
dirty
0.6.1-133-g2906152b-dirty
0.6.1-131-g82d7dda9-dirty
0.6.1-83-gd48be2c7-dirty
0.6.0-232-g298e3ddf-dirty
0.6.0-231-gb8e4f716-dirty
0.6.0-80-gcd12203b-dirty
0.6.0-75-g201cec8b-dirty
0.6.0.beta2-121-g0baae423-dirty
0.6.0.beta2-82-gefc5f15a-dirty
0.6.0.beta2-21-gc982e1b-dirty
0.6.0.beta2-20-g5fc4c58-dirty
0.6.0.beta2-17-g08fa4cb-dirty
0.5.0-291-gcc0bb21-dirty
```

**Additional Info: Old images**
Find old images of `keptn/helm-services` that ocntain the word `feature` in their tag: https://hub.docker.com/repository/registry-1.docker.io/keptn/helm-service/tags?page=1&ordering=-last_updated&name=feature

Programmatically

```
TOKEN=$(curl -s -H "Content-Type: application/json" -X POST -d '{"username": "'${DOCKER_HUB_USERNAME}'", "password": "'${DOCKER_HUB_PASSWORD}'"}' https://hub.docker.com/v2/users/login/ | jq -r .token)

curl -s -H "Authorization: JWT ${TOKEN}" "https://hub.docker.com/v2/repositories/keptn/helm-service/tags/?name=feature&ordering=-last_updated&page_size=100" | jq -r '.results|.[]|.name'
```
**Though in addition, we also need to filte rthe returned list of images by their "last updated" date, and make sure we only delete images that are older than 1 month or so.

**Definition of Done:**
- [ ] Travis CI or Github Actions is extended by a script/workflow that takes care of keeping DockerHub clean of dirty images
- [ ] Travis CI or Github Actions is extended by a script/workflow that takes care of old images that do not contain a release-tag

### PR DESCRIPTION
**Pull request by bacherfl**
Closes #1037 

Note: I could not find a possibility to create and distinguish between different cron jobs for the same branch in Travis (apparently there was a feature request for that scenario that got closed: https://github.com/travis-ci/travis-ci/issues/8820).

This means that the docker image cleanup (which takes quite some time) would be part of the daily cron job, as opposed to being executed weekly. 
@christian-kreuzberger-dtx @warber what's your opinion - should we keep that task within travis or should we discuss other approaches?
Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

### CODE REVIEW

**Comment by christian-kreuzberger-dtx** [comment]
>  @christian-kreuzberger-dtx @warber what's your opinion - should we keep that task within travis or should we discuss other approaches?

Let's move it to a GH Action with cron: https://www.jeffgeerling.com/blog/2020/running-github-actions-workflow-on-schedule-and-other-events

**Comment by christian-kreuzberger-dtx** [comment]
I just tried this out locally, am I correct in saying that we are only deleting images that are older than 30 days (controlled by `MAX_AGE=30`) ?

**Comment by bacherfl** [comment]
> I just tried this out locally, am I correct in saying that we are only deleting images that are older than 30 days (controlled by `MAX_AGE=30`) ?

Correct - we can change that limit of course, 30 days is maybe a bit high for feature/bug/dirty images

**Comment by christian-kreuzberger-dtx** [comment]
Thanks.

I just found one more thing: Deleting a tag doesn't work with an access key (it requires the actual password of the user): This is a limitation that has also been found by others, see https://github.com/docker/hub-feedback/issues/2006

So far the only known mitigation is to use an actual dockerhub account.

**Comment by christian-kreuzberger-dtx** [comment]
@bacherfl I've clicked "Ready for review" as I believe we can merge this

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

**Comment by johannes-b** [comment]
Deleting old images containing `feature`, `dirty`, `patch`, and `bugfix` is already great for cleaning up the DockerHub repos. 
However, we still have a lot of (unused) images in the repos. 
Should I create a follow-up issue to delete all images:
- starting with `201901**` -- `202006**` (< older than 6 month)
- containing a commit ID older than 6 month

File: gh-actions-scripts/cleanup_docker_images.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+# - REGISTRY_USER                                                #
+# - REGISTRY_PASSWORD                                            #
+##################################################################
+
+##################################################################
+# Configuration                                                  #
+##################################################################
+
+# list all images that should be checked
+IMAGES=("api" "bridge2" "configuration-service" "openshift-route-service" "distributor" "gatekeeper-service" "helm-service" "jmeter-service" "lighthouse-service" "mongodb-datastore" "remediation-service" "shipyard-controller" "shipyard-service")
```
[the remark this answers is not in the document]
Comment (christian-kreuzberger-dtx) [reply gh-actions-scripts/cleanup_docker_images.sh:15]: I think we can clean up those repos by manually running the script locally, we don't need to add it to the automation.

[the remark this answers is not in the document]
Comment (bacherfl) [reply gh-actions-scripts/cleanup_docker_images.sh:15]: I agree with @christian-kreuzberger-dtx - we can execute the script for those services once locally, and afterwards it should not be necessary anymore as we won't push new tags for those.

**Review by warber — APPROVED** [review]
nice, LGTM

**Pull request merged**